File: tests/tamm/Test_SVD.cpp
#include <tamm/eigen_utils.hpp>
#include <tamm/tamm.hpp>
#include <tamm/tamm_git.hpp>

#include <iomanip>

using namespace tamm;

// Reconstruction test for tamm::svd.
// Builds a random distributed A (M x N, M>=N), computes A = U diag(S) Vh, and checks
// || A - U diag(S) Vh ||_F / ||A||_F is small along with orthonormality of U and Vh.
template<typename T>
void test_svd(ExecutionContext& ec, size_t M, size_t N, Tile tilesize, const std::string& label) {
  // exhw=GPU when built with USE_CUDA/HIP/DPCPP, else CPU
  Scheduler   sch{ec};
  ExecutionHW exhw = ec.exhw();

  // EXPECTS(M >= N);

  TiledIndexSpace MR{IndexSpace{range(M)}, tilesize};
  TiledIndexSpace NC{IndexSpace{range(N)}, tilesize};

  Tensor<T> A{MR, NC};
  sch.allocate(A).execute();
  random_ip(A);

  SVDOptions opts;
  opts.full_matrices                    = true;
  auto t1                               = std::chrono::high_resolution_clock::now();
  auto [U, S, Vh]                       = svd(ec, A, opts, exhw);
  auto                          t2      = std::chrono::high_resolution_clock::now();
  std::chrono::duration<double> elapsed = t2 - t1;
  if(ec.print()) {
    std::cout << "SVD elapsed time: " << std::fixed << std::setprecision(2) << elapsed.count()
              << " seconds" << std::endl;
  }

  // Verify from the ACTUAL returned factor shapes, so this works for both full_matrices
  // (U: M x M, Vh: N x N) and reduced (U: M x K, Vh: K x N), and for M<N. Sigma is uc x vr with
  // the min(uc, vr) singular values on the diagonal; the reconstruction Ue*Sigma*Ve is then M x N
  // in every case. U's columns and Vh's rows are orthonormal in every case.
  using EM              = Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>;
  EM                 Ae = tamm_to_eigen_matrix(A);  // M x N
  EM                 Ue = tamm_to_eigen_matrix(U);  // M x uc
  EM                 Ve = tamm_to_eigen_matrix(Vh); // vr x N
  const Eigen::Index uc = Ue.cols();
  const Eigen::Index vr = Ve.rows();
  EM                 Sd = EM::Zero(uc, vr);
  const Eigen::Index Kd = std::min(uc, vr);
  for(Eigen::Index k = 0; k < Kd; k++) Sd(k, k) = S[k];

  // adjoint() (conjugate transpose) rather than transpose(): U/Vh are unitary for complex T,
  // orthogonal for real T, and adjoint() reduces to transpose() in the real case.
  const double rel = (Ue * Sd * Ve - Ae).norm() / (Ae.norm() > 0 ? Ae.norm() : 1.0);
  const double utu = (Ue.adjoint() * Ue - EM::Identity(uc, uc)).norm();
  const double vvt = (Ve * Ve.adjoint() - EM::Identity(vr, vr)).norm();

  const EM     A_reconstructed = Ue * Sd * Ve;
  const double err             = (Ae - A_reconstructed).norm();

  if(ec.print()) {
    std::cout << "SVD [" << label << "] M=" << M << " N=" << N << " tile=" << tilesize
              << " : ||A-USVh||/||A||=" << rel << " ||U^HU-I||=" << utu << " ||VhVh^H-I||=" << vvt
              << std::endl
              << "U Dims=" << Ue.rows() << "x" << Ue.cols() << "\nVt Dims=" << Ve.rows() << "x"
              << Ve.cols() << std::endl
              << "Reconstruction error: " << std::scientific << std::setprecision(3) << err
              << std::endl;
  }

  EXPECTS(rel < 1e-8);
  EXPECTS(utu < 1e-8);
  EXPECTS(vvt < 1e-8);

  sch.deallocate(A, U, Vh).execute();
}

int main(int argc, char* argv[]) {
  tamm::initialize(argc, argv);

  size_t M = 100, N = 20;
  Tile   tile = 50;
  if(argc >= 3) {
    M = std::atoi(argv[1]);
    N = std::atoi(argv[2]);
  }
  if(argc >= 4) tile = std::atoi(argv[3]);

  ProcGroup        pg = ProcGroup::create_world_coll();
  ExecutionContext ec{pg, DistributionKind::nw, MemoryManagerKind::ga};

  if(ec.print()) {
    std::cout << tamm_git_info() << std::endl;
    auto current_time   = std::chrono::system_clock::now();
    auto current_time_t = std::chrono::system_clock::to_time_t(current_time);
    auto cur_local_time = localtime(&current_time_t);
    std::cout << std::endl << "date: " << std::put_time(cur_local_time, "%c") << std::endl;

    std::cout << "nnodes: " << ec.nnodes() << ", ";
    std::cout << "nproc_per_node: " << ec.ppn() << ", ";
    std::cout << "nproc_total: " << ec.nnodes() * ec.ppn() << ", ";
    if(ec.has_gpu()) {
      std::cout << "ngpus_per_node: " << ec.gpn() << ", ";
      std::cout << "ngpus_total: " << ec.nnodes() * ec.gpn() << std::endl;
    }
    std::cout << std::endl;
    ec.print_mem_info();
    std::cout << std::endl;
  }

  test_svd<double>(ec, M, N, tile, "double");
  test_svd<std::complex<double>>(ec, M, N, tile, "complex<double>");

  tamm::finalize();
  return 0;
}

File: tests/python/Test_SVD.py
# Test_SVD.py -- Python equivalent of tests/tamm/Test_SVD.cpp
#
# Reconstruction test for tamm.svd. Builds a random distributed A (M x N), computes
# A = U diag(S) Vh, and checks ||A - U diag(S) Vh||_F / ||A||_F is small along with
# orthonormality of U and Vh.

import sys
import time

import numpy as np
import pytamm as tamm


def expects(condition, message="EXPECTS failed"):
    if not condition:
        raise AssertionError(message)


def test_svd(ec, M, N, tilesize, complex_type=False):
    sch = tamm.Scheduler(ec)
    # exhw=GPU when built with USE_CUDA/HIP/DPCPP, else CPU
    exhw = ec.exhw()

    # expects(M >= N)

    MR = tamm.TiledIndexSpace(tamm.IndexSpace(tamm.range(M)), tilesize)
    NC = tamm.TiledIndexSpace(tamm.IndexSpace(tamm.range(N)), tilesize)

    label = "complex<double>" if complex_type else "double"
    A = tamm.TensorComplexDouble([MR, NC]) if complex_type else tamm.TensorDouble([MR, NC])
    sch.allocate(A).execute()
    tamm.random_ip(A)

    opts = tamm.SVDOptions()
    opts.full_matrices = True
    t1 = time.perf_counter()
    U, S, Vh = tamm.svd(ec, A, opts, exhw)
    t2 = time.perf_counter()
    if ec.print():
        print(f"SVD elapsed time: {t2 - t1:.2f} seconds")
    # S is always real-valued, even for complex_type, matching tamm::svd/LAPACK's gesvd.
    S = np.asarray(S, dtype=np.float64)

    # Verify from the ACTUAL returned factor shapes, so this works for both full_matrices
    # (U: M x M, Vh: N x N) and reduced (U: M x K, Vh: K x N), and for M<N. Sigma is uc x vr with
    # the min(uc, vr) singular values on the diagonal; the reconstruction Ue @ Sigma @ Ve is then
    # M x N in every case. U's columns and Vh's rows are orthonormal in every case.
    Ae = A.to_numpy()   # M x N
    Ue = U.to_numpy()   # M x uc
    Ve = Vh.to_numpy()  # vr x N
    uc = Ue.shape[1]
    vr = Ve.shape[0]
    Sd = np.zeros((uc, vr), dtype=Ae.dtype)
    Kd = min(uc, vr)
    for k in range(Kd):
        Sd[k, k] = S[k]

    # conj().T (adjoint) rather than .T: U/Vh are unitary for complex_type, orthogonal for
    # real, and adjoint reduces to plain transpose in the real case.
    anorm = np.linalg.norm(Ae)
    rel = np.linalg.norm(Ue @ Sd @ Ve - Ae) / (anorm if anorm > 0 else 1.0)
    utu = np.linalg.norm(Ue.conj().T @ Ue - np.eye(uc))
    vvt = np.linalg.norm(Ve @ Ve.conj().T - np.eye(vr))

    A_reconstructed = Ue @ Sd @ Ve
    err = np.linalg.norm(Ae - A_reconstructed)

    if ec.print():
        print(f"SVD [{label}] M={M} N={N} tile={tilesize} : ||A-USVh||/||A||={rel} "
              f"||U^HU-I||={utu} ||VhVh^H-I||={vvt}")
        print(f"U Dims={Ue.shape[0]}x{Ue.shape[1]}\nVt Dims={Ve.shape[0]}x{Ve.shape[1]}")
        print(f"Reconstruction error: {err:.3e}")

    expects(rel < 1e-8)
    expects(utu < 1e-8)
    expects(vvt < 1e-8)

    sch.deallocate(A, U, Vh).execute()


def main(argv=None):
    if argv is None:
        argv = sys.argv

    tamm.initialize(argv)
    pg = tamm.ProcGroup.create_world_coll()
    ec = tamm.ExecutionContext(pg, tamm.DistributionKind.nw, tamm.MemoryManagerKind.ga)

    M, N = 100, 20
    tile = 50
    if len(argv) >= 3:
        M = int(argv[1])
        N = int(argv[2])
    if len(argv) >= 4:
        tile = int(argv[3])

    test_svd(ec, M, N, tile, complex_type=False)
    test_svd(ec, M, N, tile, complex_type=True)

    tamm.finalize()


if __name__ == "__main__":
    main()
